packages/flutter_tools/lib/src/android/java.dart

```diff
@@ -0,0 +1,245 @@
+  /// If you need to invoke the binary directly, consider adding a new method
+  /// to this class instead.
+  final String binary;
+
+  final Logger _logger;
+  final FileSystem _fileSystem;
+  final OperatingSystemUtils _os;
+  final Platform _platform;
+  final ProcessManager _processManager;
+  final ProcessUtils _processUtils;
+
+  JavaVersion? _version;
+
+  /// Returns an environment variable map with
+  /// 1. JAVA_HOME set if this object has a known home directory, and
+  /// 2. The java binary folder appended onto PATH, if the binary location is known.
+  ///
+  /// This map should be used as the environment when invoking any Java-dependent
+  /// processes, such as Gradle or Android SDK tools (avdmanager, sdkmanager, etc.)
+  Map<String, String> getJavaEnvironment() {
+    return <String, String>{
+      if (home != null) _javaHomeEnvironmentVariable: home!,
+      'PATH': _fileSystem.path.dirname(binary) +
+                        _os.pathVarSeparator +
+                        _platform.environment['PATH']!,
+    };
+  }
+
+  /// Returns the version of java in the format \d(.\d)+(.\d)+
+  /// Returns null if version could not be determined.
+  JavaVersion? getVersion() {
+    if (_version == null) {
+      final RunResult result = _processUtils.runSync(
+        <String>[binary, '--version'],
+        environment: getJavaEnvironment(),
+      );
+      if (result.exitCode != 0) {
+        _logger.printTrace('java --version failed: exitCode: ${result.exitCode}'
+          ' stdout: ${result.stdout} stderr: ${result.stderr}');
+      }
+      _version = JavaVersion.tryParseFromJavaOutput(result.stdout, logger: _logger);
+    }
+
+    return _version;
+  }
+
+  bool canRun() {
+    return _processManager.canRun(binary);
+  }
+}
+
+String? _findJavaHome({
+  required Logger logger,
+  required AndroidStudio? androidStudio,
+  required Platform platform,
+}) {
+  final String? androidStudioJavaPath = androidStudio?.javaPath;
+  if (androidStudioJavaPath != null) {
+    logger.printTrace("Using Android Studio's java.");
+    return androidStudioJavaPath;
+  }
+
+  final String? javaHomeEnv = platform.environment[_javaHomeEnvironmentVariable];
+  if (javaHomeEnv != null) {
+    logger.printTrace('Using JAVA_HOME from environment valuables.');
+    return javaHomeEnv;
+  }
+  return null;
+}
+
+String? _findJavaBinary({
+  required Logger logger,
+  required String? javaHome,
+  required FileSystem fileSystem,
+  required OperatingSystemUtils operatingSystemUtils,
+  required Platform platform,
+}) {
+  try {
+    if (javaHome != null) {
+      return fileSystem.path.join(javaHome, 'bin', 'java');
+    }
+
+    // Fallback to PATH based lookup.
+    final String? pathJava = operatingSystemUtils.which(_kJavaExecutable)?.path;
+    if (pathJava != null) {
+      logger.printTrace('Using java from PATH.');
+    } else {
+      logger.printTrace('Could not find java path.');
+    }
+    return pathJava;
+  } on Exception catch (e) {
```

Comment:
exactly what exceptions would you expect to be safe for us to ignore? and can we just explicitly catch those. my concern is I update this code and break this, we could be returning null 100% of the time and no one would know.

Reply:
Even after looking back at current code in `master`, I can't remember why I added this try-catch here, so I removed it.